### PR TITLE
WIP: Automatically retry when encounter connection reset by peer error from aws api

### PR DESCRIPTION
**Pull request by jiashuChen**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #10715 

Draft PR, I'm not sure if this is the most correct way to solve this issue, raising as a draft, feedback will be appreciated. The goal of this pull request is to resolve to have a global wrapper function to deal with connection reset by peer error from aws api, which seems to impact multiple aws region and services.

```
│ Error: error deleting IAM Role (IAM-ROLE-NAME): RequestError: send request failed
│ caused by: Post "https://iam.amazonaws.com/": read tcp IP:PORT->DIFFERENT_IP:PORT: read: connection reset by peer
```

Furthermore, this helps resolve the issue with `resource.Retry` being marked deprecated.

https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/resource/wait.go#L66-L69

Lastly, this pull request now seems to have more than 100 labels automatically applied to it, which has caused verification to fail. I'm not sure how to resolve it. Some help will be appreciated.

NOTE: although there are many files changes, the core of the change is here in this file:
https://github.com/hashicorp/terraform-provider-aws/blob/03ea02c693ab28c09cbcce39dfe5ba98676eb33a/aws/internal/tfresource/retry.go#L119-L129

### CODE REVIEW

File: aws/internal/tfresource/retry.go [hunk 1]
```diff
@@ -114,3 +115,15 @@ func RetryConfigContext(ctx context.Context, delay time.Duration, delayRand time
 	// more likely to be useful
 	return resultErr
 }
+
+func RetryOnConnectionResetByPeer(timeout time.Duration, f resource.RetryFunc) error {
```
Comment (jiashuChen) [inline aws/internal/tfresource/retry.go:119]: Main Change

File: aws/internal/tfresource/retry_test.go [hunk 2]
```diff
@@ -97,3 +98,55 @@ func TestRetryConfigContext_error(t *testing.T) {
 		t.Fatal("timeout")
 	}
 }
+
+func TestRetryOnConnectionResetByPeer(t *testing.T) {
```
Comment (jiashuChen) [inline aws/internal/tfresource/retry_test.go:102]: Unit Test

**Comment by AdamTylerLynch** [comment]
Question for the author: the AWS GoSDK has retry configured by default for all client operations. The retry mechanism will retry 3 times by default and use exponential back off with jitter.

I’m curious as to how this PR will be any different? I would suggest modifying the default policy for the SDK rather than implement additional retry logic, specifically because that logic does not include exponential backoff with jitter.

https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/retries-timeouts/

Additionally, TCP reset could be happening for a number of reasons. I use Terraform all the time, running the Acceptance Tests, 20 threads at a time, and I’ve never gotten a TCP reset error. 

Have you checked your network route to AWS? A proxy, machine-in-the-middle, or AWS API throttling are all potential sources of TCP reset. I would recommend checking your CloudTrail logs for your account to see if you are being throttled.

**Comment by jiashuChen** [comment]
> Question for the author: the AWS GoSDK has retry configured by default for all client operations. The retry mechanism will retry 3 times by default and use exponential back off with jitter.
> 
> I’m curious as to how this PR will be any different? I would suggest modifying the default policy for the SDK rather than implement additional retry logic, specifically because that logic does not include exponential backoff with jitter.
> 
> https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/retries-timeouts/
> 
> Additionally, TCP reset could be happening for a number of reasons. I use Terraform all the time, running the Acceptance Tests, 20 threads at a time, and I’ve never gotten a TCP reset error.
> 
> Have you checked your network route to AWS? A proxy, machine-in-the-middle, or AWS API throttling are all potential sources of TCP reset. I would recommend checking your CloudTrail logs for your account to see if you are being throttled.

Thanks. I'll close the current pull requests.

**Pull request closed without merging**